core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/ChecksumCalculatingAsyncRequestBody.java

```diff
@@ -243,4 +248,35 @@ private Iterable<ByteBuffer> buffer(ByteBuffer bytes) {
         }
     }
 
+    public static class OnNextGuaranteedSubscriber extends DelegatingSubscriber<ByteBuffer, ByteBuffer> {
+
+        private volatile boolean onNextInvoked;
+
+        public OnNextGuaranteedSubscriber(Subscriber<? super ByteBuffer> subscriber) {
+            super(subscriber);
+        }
+
+        @Override //needed?
+        public void onSubscribe(Subscription subscription) {
+            super.onSubscribe(subscription);
+        }
+
+        @Override
+        public void onNext(ByteBuffer t) {
+            if (!onNextInvoked) {
+                onNextInvoked = true;
+            }
+
+            subscriber.onNext(t);
+        }
+
+        @Override
+        public void onComplete() {
+            if (!onNextInvoked) {
+                subscriber.onNext(ByteBuffer.wrap(new byte[0]));
```

Comment:
Just a note: request demand should be handled in FlatteningSubscriber as part of `flatMapIterable`